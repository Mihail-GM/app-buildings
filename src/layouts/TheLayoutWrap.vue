<template>
	<div>
		<v-main class="mt-5">
			<v-container>
				<slot></slot>
			</v-container>
		</v-main>

		<v-navigation-drawer v-model="drawer" app>
			<v-list>
				<v-list-item
					v-for="item in menuItems"
					:key="item.title"
					:to="item.path"
					:id="'mobile-navigation-' + item.title"
				>
					<v-list-item-action>
						<v-icon>{{ item.icon }}</v-icon>
					</v-list-item-action>
					<v-list-item-content>{{ item.title }}</v-list-item-content>
				</v-list-item>
			</v-list>
		</v-navigation-drawer>

		<v-toolbar dark bottom width="100%" class="px-md-10 px-sm-0 px-xs-0" style="position: fixed;">
			<span class="hidden-md-and-up">
		        <v-app-bar-nav-icon @click="drawer = !drawer"/>
	        </span>

			<v-btn
				class="hidden-sm-and-down"
				text
				x-large
				v-for="item in menuItems"
				:key="item.title"
				:id="'navigation-' + item.title.toLowerCase()"
				@click.stop="$router.push(item.path)"
			>
				<v-icon left dark> {{ item.icon }}</v-icon>
				{{ item.title }}
			</v-btn>

			<v-spacer></v-spacer>

			<v-toolbar-title>
				<h3
					id="app-title-navigation"
					@click.stop="$router.push(`/`)"
					style="cursor: pointer"
				>
					{{ appTitle }}
				</h3>
			</v-toolbar-title>

		</v-toolbar>
	</div>
</template>

<script>

	export default {
		name: "TheLayoutWrap",

		data: () => ({
			drawer: false,
			appTitle: 'Buildings App',
			menuItems: [
				{
					icon: 'mdi-home-circle',
					title: 'Home',
					path: '/'
				},
				{
					icon: 'mdi-office-building',
					title: 'Buildings',
					path: '/buildings'
				},
			]
		}),
	};
</script>

<style scoped>

</style>