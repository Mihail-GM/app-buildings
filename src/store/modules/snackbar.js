const state = {
	snackbarMessageData: {
		message: '',
		duration: 3000,
		mode: 'success'
	},
};

const getters = {
	getSnackbarMessage(state) {
		return state.snackbarMessageData;
	},
};

const mutations = {
	setSnackbarMessage(state, {message, duration = 4000, mode = 'success', buttonText = 'Close'}) {
		state.snackbarMessageData = {message, duration, mode, buttonText};
	},
}

const actions = {
	showSnackbarMessage(vuexContext, messageData) {
		vuexContext.commit('setSnackbarMessage', messageData);
	},
}

export default {
	namespaced: true,
	state,
	getters,
	mutations,
	actions
}