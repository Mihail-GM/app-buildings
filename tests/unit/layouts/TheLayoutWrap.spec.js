import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import {merge} from "lodash";

import TheLayoutWrap from "@/layouts/TheLayoutWrap";
import snackbar from "@/store/modules/snackbar";


describe('TheLayoutWrap.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify

	beforeEach(() => {
		vuetify = new Vuetify()
	})

	function createStore(overrides) {
		let snackbarMessageData = {
			message: 'Test',
			duration: 3000,
			mode: 'success'
		};

		const defaultStoreConfig = {
			modules: {
				snackbar: {
					namespaced: true,
					state: {
						snackbarMessageData
					},
					getters: snackbar.getters,
					mutations: snackbar.mutations,
					actions: snackbar.actions,
				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),

			mocks: {
				$route: {
					path: '/',
				},
				// Vue Router
				$router: {
					push: jest.fn(),
				},
			},
		}

		const dataOptions = {
			data() {
				return {
					drawer: false,
					appTitle: 'Buildings App',
					menuItems: [
						{
							icon: 'mdi-home-circle',
							title: 'Home',
							path: '/'
						},
						{
							icon: 'mdi-home-circle',
							title: 'Buildings',
							path: '/buildings'
						},
					]
				}
			}
		}

		return mount(TheLayoutWrap, merge(defaultMountingOptions, dataOptions, overrides))
	}

	it('should render component ', () => {
		const wrapper = createWrapper()

		expect(wrapper.is(TheLayoutWrap)).toBe(true);
	})

	it('app title link should redirect to home "/"', () => {
		const wrapper = createWrapper();

		const spy = jest.fn();
		wrapper.vm.$router.push = spy;

		let homeButton = wrapper.find('#app-title-navigation')
		homeButton.trigger('click');

		expect(spy).toHaveBeenCalledWith('/');
	})

	it('should render first item on mobile menu with correct link', async () => {
		const wrapper = createWrapper();

		let menuLink = wrapper.findAll({name: 'v-list-item'}).at(0);

		expect(menuLink.props().to).toBe('/');
	})

	it('should render second item on mobile menu with correct link', async () => {
		const wrapper = createWrapper();

		let menuLink = wrapper.findAll({name: 'v-list-item'}).at(1);

		expect(menuLink.props().to).toBe('/buildings');
	})


	it('should open /buildings page when click on buildings button on the bottom toolbar', async () => {
		const wrapper = createWrapper();
		const spy = jest.fn();
		wrapper.vm.$router.push = spy;

		let button = wrapper.find('#navigation-buildings')
		button.trigger('click');

		expect(spy).toHaveBeenCalledWith('/buildings');
	})

	it('should open / page when click on home button on the bottom toolbar', async () => {
		const wrapper = createWrapper();
		const spy = jest.fn();
		wrapper.vm.$router.push = spy;

		let button = wrapper.find('#navigation-home')
		button.trigger('click');

		expect(spy).toHaveBeenCalledWith('/');
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
