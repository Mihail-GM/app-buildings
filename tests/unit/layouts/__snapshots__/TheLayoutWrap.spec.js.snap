// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheLayoutWrap.vue should match snapshot 1`] = `
<div>
  <main class="v-main mt-5" style="padding: 0px 0px 0px 0px;">
    <div class="v-main__wrap">
      <div class="container"></div>
    </div>
  </main>
  <nav class="v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile theme--light" style="height: 100vh; top: 0px; transform: translateX(-100%); width: 256px;">
    <div class="v-navigation-drawer__content">
      <div class="v-list v-sheet theme--light">
        <router-link tabindex="0" id="mobile-navigation-Home" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__action"><i aria-hidden="true" class="v-icon notranslate mdi mdi-home-circle theme--light"></i></div>
          <div class="v-list-item__content">Home</div>
        </router-link>
        <router-link tabindex="0" id="mobile-navigation-Buildings" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__action"><i aria-hidden="true" class="v-icon notranslate mdi mdi-home-circle theme--light"></i></div>
          <div class="v-list-item__content">Buildings</div>
        </router-link>
      </div>
    </div>
    <div class="v-navigation-drawer__border"></div>
  </nav>
  <header class="px-md-10 px-sm-0 px-xs-0 v-sheet theme--dark v-toolbar v-toolbar--bottom" style="width: 100%; height: 64px; position: fixed;">
    <div class="v-toolbar__content" style="height: 64px;"><span class="hidden-md-and-up"><button type="button" class="v-app-bar__nav-icon v-btn v-btn--icon v-btn--round theme--dark v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu theme--dark"></i></span></button></span> <button type="button" class="hidden-sm-and-down v-btn v-btn--text theme--dark v-size--x-large" id="navigation-home"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-home-circle theme--dark"></i>
			Home
		</span></button><button type="button" class="hidden-sm-and-down v-btn v-btn--text theme--dark v-size--x-large" id="navigation-buildings"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-home-circle theme--dark"></i>
			Buildings
		</span></button>
      <div class="spacer"></div>
      <div class="v-toolbar__title">
        <h3 id="app-title-navigation" style="cursor: pointer;">
          Buildings App
        </h3>
      </div>
    </div>
  </header>
</div>
`;
